src/main/java/network/brightspots/rcv/CandidateRankingsList.java

```diff
@@ -0,0 +1,121 @@
+ * Design: Assumes valid ballots but supports overvotes and skipped rankings.
+ * Conditions: Always.
+ * Version history: see https://github.com/BrightSpots/rcv.
+ */
+
+
+package network.brightspots.rcv;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import javafx.util.Pair;
+
+class CandidateRankingsList implements Iterable<Pair<Integer, CandidatesAtRanking>> {
+  class CandidateRankingsListIterator implements Iterator<Pair<Integer, CandidatesAtRanking>> {
+    private int iteratorIndex = 0;
+
+    private CandidateRankingsList list;
+
+    CandidateRankingsListIterator(CandidateRankingsList list) {
+      this.list = list;
+    }
+
+    public boolean hasNext() {
+      if (iteratorIndex < list.rankings.length) {
+        return true;
+      }
+      return false;
+    }
+
+    public Pair<Integer, CandidatesAtRanking> next() {
+      if (iteratorIndex == list.rankings.length) {
+        throw new NoSuchElementException();
+      }
+
+      iteratorIndex++;
+      while (!hasRankingAt(iteratorIndex)) {
+        iteratorIndex++;
+        assert iteratorIndex <= list.rankings.length;
+      }
+
+      // Note: round numbers are 1-indexed externally, 0-indexed internally,
+      // thus why we return a different value than what we index into here
+      return new Pair<>(iteratorIndex, list.rankings[iteratorIndex - 1]);
+    }
+
+    public void remove() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  private CandidatesAtRanking[] rankings;
+  private int numRankings;
+  private int iteratorIndex = 0;
+
+  CandidateRankingsList(List<Pair<Integer, String>> rawRankings) {
+    if (rawRankings.isEmpty()) {
+      this.rankings = new CandidatesAtRanking[0];
+      numRankings = 0;
+      return;
+    }
+
+    rawRankings.sort(Comparator.comparingInt(Pair::getKey));
+
+    // Initialize up to maxRankings, leaving empty arrays for any undervotes
+    int minRanking = rawRankings.get(0).getKey();
+    if (minRanking <= 0) {
+      Logger.severe("Rankings must start at 1, but you have a ranking at %d", minRanking);
+      // TODO should I throw from here or not?
+    }
+
+    int maxRanking = rawRankings.get(rawRankings.size() - 1).getKey();
+    this.rankings = new CandidatesAtRanking[maxRanking];
+    for (int i = 0; i < maxRanking; ++i) {
+      this.rankings[i] = new CandidatesAtRanking();
+    }
+
+    // Populate, which is most efficient in the common case of no overvotes
+    for (Pair<Integer, String> ranking : rawRankings) {
+      this.rankings[ranking.getKey() - 1].addCandidate(ranking.getValue());
+    }
+
+    for (CandidatesAtRanking c : this.rankings) {
+      if (c.count() != 0) {
+        ++numRankings;
+      }
+    }
+  }
+
+  CandidatesAtRanking get(int i) {
+    return rankings[i - 1];
+  }
+
+  boolean hasRankingAt(int num) {
+    return num >= 1 && num <= rankings.length && rankings[num - 1].count() != 0;
```

Comment:
Not sure I follow -- you mean, just return `rankings[num - 1].count() != 0` and don't check if `num` is in bounds?

I think the caller should at least be able to have `num > rankings.length`, since a natural call may be `while(++i) hasRankingAt(i)` -- i.e. don't check against `maxNumRankings`. I'm fine asserting that `num >= 1` though, since a caller should never call with a zero or negative index.